src/coreclr/vm/reflectioninvocation.cpp

```diff
@@ -1243,6 +1227,77 @@ lExit: ;
 }
 FCIMPLEND
 
+FCIMPL1(FC_BOOL_RET, RuntimeFieldHandle::IsFastPathSupported, ReflectFieldObject *pFieldUNSAFE)
+{
+    CONTRACTL {
+        FCALL_CHECK;
+    }
+    CONTRACTL_END;
+
+    REFLECTFIELDREF refField = (REFLECTFIELDREF)ObjectToOBJECTREF(pFieldUNSAFE);
+    if (refField == NULL)
+        FCThrowRes(kArgumentNullException, W("Arg_InvalidHandle"));
+
+    FieldDesc* pFieldDesc = refField->GetField();
+    return pFieldDesc->IsThreadStatic() || pFieldDesc->IsEnCNew() ? FALSE : TRUE;
+}
+FCIMPLEND
+
+FCIMPL1(INT32, RuntimeFieldHandle::GetInstanceFieldAddress, ReflectFieldObject *pFieldUNSAFE)
+{
+    CONTRACTL {
+        FCALL_CHECK;
+        PRECONDITION(CheckPointer(pFieldUNSAFE));
+    }
+    CONTRACTL_END;
+
+    REFLECTFIELDREF refField = (REFLECTFIELDREF)ObjectToOBJECTREF(pFieldUNSAFE);
+    if (refField == NULL)
+        FCThrowRes(kArgumentNullException, W("Arg_InvalidHandle"));
+
+    FieldDesc* pFieldDesc = refField->GetField();
+
+    // IsFastPathSupported needs to checked before calling this method.
+    _ASSERTE(!pFieldDesc->IsEnCNew());
+
+    return pFieldDesc->GetOffset();
+}
+FCIMPLEND
+
+FCIMPL2(void*, RuntimeFieldHandle::GetStaticFieldAddress, ReflectFieldObject *pFieldUNSAFE, CLR_BOOL *isBoxed)
+{
+    CONTRACTL {
+        FCALL_CHECK;
+        PRECONDITION(CheckPointer(pFieldUNSAFE));
+    }
+    CONTRACTL_END;
+
+    REFLECTFIELDREF refField = (REFLECTFIELDREF)ObjectToOBJECTREF(pFieldUNSAFE);
+    if (refField == NULL)
+        FCThrowRes(kArgumentNullException, W("Arg_InvalidHandle"));
+
+    void *ret = NULL;
+    FieldDesc* pFieldDesc = refField->GetField();
+
+    // IsFastPathSupported needs to checked before calling this method.
+    _ASSERTE(!pFieldDesc->IsEnCNew());
+    _ASSERTE(!refField->GetField()->IsThreadStatic());
```

Comment:
Addresses of regular static fields in collectitble assemblies are allocated on the GC heap and can move.

This needs to have similar set of cases as the JIT: https://github.com/dotnet/runtime/blob/1e8f741f3328ef5cde6129476c279f5aef450b66/src/coreclr/vm/jitinterface.cpp#L1563-L1571